### PR TITLE
Support upload for Azure artifacts

### PR DESCRIPTION
**Pull request by lukebjerring**
## Description
Works towards https://github.com/web-platform-tests/wpt.fyi/issues/714

Unfortunately, the only API for an artifact is a zip file, with the report nested inside. This PR aims to support `report_path` which implies this case, and extracts that path as the report file for upload.

## Review Information
This has been deployed as `checks` in staging, where I've verified it actually succeeds:
https://staging.wpt.fyi/runs?sha=9b4d522493a45f24338966d2ba9206cd3be383ff

### CODE REVIEW

File: shared/appengine.go [hunk 1]
```diff
@@ -44,6 +54,18 @@ func (a AppEngineAPIImpl) Context() context.Context {
 	return a.ctx
 }
 
+// GetHTTPClient returns an HTTP client for the current context.
+func (a AppEngineAPIImpl) GetHTTPClient() *http.Client {
```
Comment (lukebjerring) [inline shared/appengine.go]: @mdittmer - here's the client abstraction I mentioned.

File: api/checks/azure_test.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+func TestHandleAzurePipelinesEvent(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	sha := strings.Repeat("0123456789", 4)
+	detailsURL := "https://dev.azure.com/web-platform-tests/b14026b4-9423-4454-858f-bf76cf6d1faa/_build/results?buildId=123"
+	event := getCheckRunCreatedEvent("completed", "lukebjerring", sha)
+	event.CheckRun.DetailsURL = &detailsURL
+
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```
Comment (lukebjerring) [inline api/checks/azure_test.go]: @mdittmer - here's a "fake" http server (as opposed to a mock one). The calling code just expects "given this http.Client, and this URL, fetch it" so we mock the URL we give out (as the server's URL) and the client we give out (as the server's Client).

File: api/checks/api.go [hunk 3]
```diff
@@ -172,6 +179,61 @@ func (s checksAPIImpl) CreateWPTCheckSuite(appID, installationID int64, sha stri
 	return suite != nil, err
 }
 
+func (s checksAPIImpl) GetAzureArtifactsURL(owner, repo string, buildID int64) string {
```
Comment (Hexcles) [inline api/checks/api.go]: I'm confused by `golint`. Why doesn't it complain the lack of comment for this exported method?

File: api/README.md [hunk 4]
```diff
@@ -355,7 +355,10 @@ __`callback_url`__: (Optional) A URL that the processor should `POST` when succe
 create the TestRun. Defaults to /api/results/create in the current project's environment (e.g. wpt.fyi for
 wptdashboard, staging.wpt.fyi for wptdashboard-staging).
 
-__`result_file`__: A **gzipped** JSON file produced by `wpt run --log-wptreport`.
+__`result_file`__: A **gzipped** JSON file produced by `wpt run --log-wptreport`, or a zip file containing
+the report. If the file is a zip, `report_path` **must** be specified.
+
+__`report_path`__: (Optional) Path to the report file within the result file, if applicable.
```
Comment (Hexcles) [inline api/README.md]: This change needs to be reverted.

File: api/receiver/receive_results.go [hunk 5]
```diff
@@ -45,15 +45,14 @@ func HandleResultsUpload(a AppEngineAPI, w http.ResponseWriter, r *http.Request)
 		uploader = username
 	}
 
-	// Most form methods (e.g. FormValue, FormFile) will call
+	// Most form methods (e.g. PostFormValue, FormFile) will call
```
Comment (Hexcles) [inline api/receiver/receive_results.go]: This whole file looks like a bad merge.

File: api/receiver/receive_results_test.go [hunk 6]
```diff
@@ -128,7 +128,6 @@ func TestHandleResultsUpload_extra_params(t *testing.T) {
 		"browser_version": "",
 		"os_name":         "",
 		"os_version":      "",
-		"callback_url":    "",
```
Comment (Hexcles) [inline api/receiver/receive_results_test.go]: Ditto.

**Comment by Hexcles** [comment]
Another reason why we should have a separate `api/azure` package is because we might need to run this adapter in a separate AppEngine service, because of memory constraints: when we start to run full runs on Azure (e.g. Edge on Win10), the zip->gzip transformation may take hundreds of MB of memory. If we have a separate package for the Azure adapter, we can then easily add a new entry point (URL route) to this code and run it as a webhook independent from our GitHub App (i.e. the same as the setup of our Taskcluster listener now).

File: api/receiver/receive_results_test.go (hunk 6 above)
Comment (lukebjerring) [reply api/receiver/receive_results_test.go]: Done.

File: api/receiver/receive_results.go (hunk 5 above)
Comment (lukebjerring) [reply api/receiver/receive_results.go]: Done.

File: api/README.md (hunk 4 above)
Comment (lukebjerring) [reply api/README.md]: Done.

File: api/checks/api.go (hunk 3 above)
Comment (lukebjerring) [reply api/checks/api.go]: Not sure either.

**Comment by lukebjerring** [comment]
Moved the code to `/api/azure`, PTAL.

**Comment by lukebjerring** [comment]
FWIW, I manually deployed this as version `checks`, and you can see the results here:
https://github.com/lukebjerring/wpt/pull/13/checks?sha=8d996d10a2f1ce9385921df1a1dc44f9b565d28f

File: api/azure/webhook.go [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+	if err != nil {
+		return err
+	}
+
+	fileField, err := writer.CreateFormFile("result_file", "wpt_report.json")
+	_, err = io.Copy(fileField, bytes.NewReader(data))
+	if err != nil {
+		return err
+	}
+	if err := writer.Close(); err != nil {
```
[the remark this answers is not in the document]
Comment (lukebjerring) [reply api/azure/webhook.go]: Done.

File: api/azure/webhook.go [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+	host := aeAPI.GetHostname()
+	writer.WriteField("callback_url", fmt.Sprintf("https://%s/api/results/create", host))
+
+	data, err := azureAPI.FetchAzureArtifact(artifact)
+	if err != nil {
+		return err
+	}
+
+	fileField, err := writer.CreateFormFile("result_file", "wpt_report.json")
+	_, err = io.Copy(fileField, bytes.NewReader(data))
```
[the remark this answers is not in the document]
Comment (lukebjerring) [reply api/azure/webhook.go]: Done.

**Pull request merged**